### PR TITLE
Add FundingCanceled

### PR DESCRIPTION
**Pull request by TheCharlatan**

### CODE REVIEW

File: src/swapd/runtime.rs [hunk 1]
```diff
@@ -2033,6 +2045,27 @@ impl Runtime {
...
+                            TxLabel::Lock
+                                if self
+                                    .temporal_safety
+                                    .stop_funding_before_cancel(*confirmations)
+                                    && self.state.safe_cancel()
+                                    && self.state.swap_role() == SwapRole::Alice
+                                    && self.awaiting_funding =>
+                            {
+                                warn!(
+                                    "{} | Alice, the swap may be cancelled soon. For your safety, do not fund this swap anymore",
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/swapd/runtime.rs:2057]: No, she funds into a swap that was already cancelled.

[the remark this answers is not in the document]
Comment (Lederstrumpf) [reply src/swapd/runtime.rs:2057]: right, but this means Bob's leaked his ed25519 private keyshare, so it's "safe", but superfluous.

**Review by Lederstrumpf — APPROVED** [review]
tACK @https://github.com/farcaster-project/farcaster-node/pull/356/commits/90094c00ea06067a2c33dbf1062198d5de00a488

**Comment by TheCharlatan** [comment]
Obviously should be merged after #354 .

File: src/swapd/runtime.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/swapd/runtime.rs:2059]: ACKCHYUALLY
.. you are.
I was thinking of a scenario where Alice restarts after cancel, which is yet untested, but we don't really know what happens during recovery in different stages of execution yet anyway. So your message is definitely more appropriate. I'll add it once I have to rebase on top of #354 .

**Comment by zkao** [comment]
I'm having a hard time understanding the part not related to FundingCancelled

will open a new PR with a couple added commit, reverting some stuff, moving others

**Comment by zkao** [comment]
`self.syncer_state.awaiting_funding == true` is for local node only

**Comment by zkao** [comment]
remote doesnt await funding, but then FundingCompleted is also not sent

**Comment by zkao** [comment]
i dont get the logic here

**Comment by zkao** [comment]
logic changed too much compared to what it was, before FundingCompleted was sent to both local and remote, now only to local,  so the merged #354 uses stuff that is not there anymore

**Comment by TheCharlatan** [comment]
Here is what I think my code is doing:

`FundingCompleted` is emitted when `AddressTransaction` is received, the `TxLabel` of the transaction is `Funding` and `awaiting_funding` is  `true`.
OR
`FundingCompleted` is emitted when `AddressTransaction` is received , the `swap_role` is `Alice`, the `TxLabel` of the transaction is `AccLock` and `awaiting_funding` is `true`.
OR
`FundingCanceled` is emitted when `TransactionConfirmation` is received, the `TxLabel` of the transaction is `Lock`, the `swap_role` is `Alice`, the `cancel` tx has not been seen and the `awaiting_funding` is `true`.
OR
`FundingCanceled` is emited when `TransactionConfirmation` is received, the `TxLabel` of the transaction is `Cancel`, the `swap_role` is `Alice`, the monero are not locked yet and `awaiting_funding` is `true`.

It should emit either the `FundingCompleted` or `FundingCanceled` only once per running swap. I've tested this extensively now and have not run into problems yet. Where do you see the flaw in the logic?

**Comment by zkao** [comment]
Before it was emmiting twice per running swap, you changed the logic, and it breaks things downstream

**Comment by zkao** [comment]
unreachable line: https://github.com/farcaster-project/farcaster-node/blob/78a72508207240b46df3c5ff7d2e4b26dedf12dd/src/farcasterd/runtime.rs#L1151

**Comment by TheCharlatan** [comment]
That line does not exist on my branch, where is this from?

**Comment by TheCharlatan** [comment]
> Before it was emmiting twice per running swap, you changed the logic, and it breaks things downstream

That was a bug and it only emitted twice for Bob, so it was also inconsistent. What does it break?

**Comment by zkao** [comment]
you accepted that on yesterday's PR, it was a pain to deal with that anarchism of having to deal with local and remote fundings, and now u just decided to drop it, thank you!

**Review by Lederstrumpf — APPROVED** [review]

**Pull request merged**